jetty-core/jetty-siwe/src/main/java/org/eclipse/jetty/security/siwe/EthereumAuthenticator.java

```diff
@@ -0,0 +1,831 @@
+        if (LOG.isDebugEnabled())
+            LOG.debug("Authentication FAILED: {}", message);
+
+        if (_errorPath == null)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("auth failed 403");
+            if (response != null)
+                Response.writeError(request, response, callback, HttpStatus.FORBIDDEN_403, message);
+        }
+        else
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("auth failed {}", _errorPath);
+
+            String contextPath = Request.getContextPath(request);
+            String redirectUri = URIUtil.addPaths(contextPath, _errorPath);
+            if (message != null)
+            {
+                String query = URIUtil.addQueries(ERROR_PARAMETER + "=" + UrlEncoded.encodeString(message), _errorQuery);
+                redirectUri = URIUtil.addPathQuery(URIUtil.addPaths(contextPath, _errorPath), query);
+            }
+
+            int redirectCode = request.getConnectionMetaData().getHttpVersion().getVersion() < HttpVersion.HTTP_1_1.getVersion()
+                ? HttpStatus.MOVED_TEMPORARILY_302 : HttpStatus.SEE_OTHER_303;
+            Response.sendRedirect(request, response, callback, redirectCode, redirectUri, true);
+        }
+    }
+
+    protected Fields getParameters(Request request)
+    {
+        try
+        {
+            Fields queryFields = Request.extractQueryParameters(request);
+            Fields formFields = FormFields.from(request).get();
+            return Fields.combine(queryFields, formFields);
+        }
+        catch (InterruptedException | ExecutionException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public boolean isLoginPage(String uri)
+    {
+        return matchURI(uri, _loginPath);
+    }
+
+    public boolean isAuthenticationRequest(String uri)
+    {
+        return matchURI(uri, _authenticationPath);
+    }
+
+    public boolean isNonceRequest(String uri)
+    {
+        return matchURI(uri, _noncePath);
+    }
+
+    private boolean matchURI(String uri, String path)
+    {
+        int jsc = uri.indexOf(path);
+        if (jsc < 0)
+            return false;
+        int e = jsc + path.length();
+        if (e == uri.length())
+            return true;
+        char c = uri.charAt(e);
+        return c == ';' || c == '#' || c == '/' || c == '?';
+    }
+
+    public boolean isErrorPage(String pathInContext)
+    {
+        if (_errorPath == null)
+            return false;
+        return pathInContext != null && (pathInContext.equals(_errorPath));
+    }
+
+    protected String createNonce(Session session)
+    {
+        String nonce = EthereumUtil.createNonce();
+        synchronized (session)
+        {
+            @SuppressWarnings("unchecked")
+            Set<String> attribute = (Set<String>)session.getAttribute(NONCE_SET_ATTR);
+            if (attribute == null)
+                session.setAttribute(NONCE_SET_ATTR, attribute = new FixedSizeSet<>(5));
```

Comment:
Why do we need the `FixedSizeSet` class? Isn't this just setting a single element?

Reply:
Probably, but I've had issues with openId where multiple authentication requests come in from the browser and then one of them fails because the nonce is forgotten and causes the other ones to fail.